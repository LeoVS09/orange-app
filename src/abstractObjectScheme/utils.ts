import { AosField, AosFieldType, AosSimpleField, AosRelationsField } from './types';

export const isSimpleAosField = (field: AosField): field is AosSimpleField => 
    !isRelationsAosField(field)

export const isRelationsAosField = (field: AosField): field is AosRelationsField => 
    field.type === AosFieldType.OneToOne || 
    field.type === AosFieldType.OneToMany
    
