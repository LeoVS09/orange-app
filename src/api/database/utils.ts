import {APIClient, makeClient} from './client'
import urls from '../urls.json'
import deepMap from 'deep-map'

// const DEBUG = process.env.NODE_ENV !== 'production'
const DEBUG = false

export const client = makeClient(urls.DATABASE_SERVER)


// TODO: complete typesation for non parametred queries
export function generateQuery<V, T, R>(query: any, formatter: (result: T) => R | null, mock?: (variables: V) => Promise<R>) {
   if (DEBUG && mock) {
      return (variables: V) => mock(variables)
   }

   return mapRequester((client) => (variables: V) =>
      client.query<T>({
         query,
         variables,
      })
         .then((result) => result.data && formatter(result.data)))
}

export function generateSimpleQuery<T, R>(query: any, formatter: (result: T) => R | null, mock?: () => Promise<R>) {
   if (DEBUG && mock) {
      return () => mock()
   }

   return mapRequester((client) => () =>
      client.query<T>({
         query,
      })
         .then((result) => result.data && formatter(result.data)),
   )
}

export function generateMutation<V, T, R>(mutation: any, formatter: (result: T) => R | null, mock?: (variables: V) => Promise<R>) {
   if (DEBUG && mock) {
      return (variables: V) => mock(variables)
   }

   return mapRequester((client) => (variables: V) =>
      client.mutate<T>({
         mutation,
         variables,
      })
         .then((result) => result.data && formatter(result.data)),
   )
}

// mapRequester transform date string to date objects, but currently TypeScript don't have support
// for so advanced templates, so just know this fact when set types of request result

function mapRequester<R>(requester: (client: APIClient) => R): R  {
   const wrapped = requester(client)

   // @ts-ignore
   return async (input: any) => {
      // @ts-ignore
      const result = await wrapped(input)
      if (!result) {
         return result
      }

      return dateToStringFormatter(result)
   }
}

const keysForMapDays: string[] = []
const keyMath = /.+(Date|At)$/gm

// Transform to date types fields with matched names
function dateToStringFormatter <T>(t: T): T {
   return deepMap<T>(t, (value, key) => {
      if (typeof key === 'number') {
         return value
      }

      if (keysForMapDays.indexOf(key) === -1 && !key.match(keyMath)) {
         return value
      }

      if (value === null || value === undefined) {
         return value
      }

      return new Date(value)
   })
}
