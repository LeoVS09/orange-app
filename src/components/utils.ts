
export function toStringWhenDefined(value: any): string {
  if (value === undefined)
    return ''

  if (value === null)
    return ''

  return `${value}`
}

export function isDate(date: any) {
  return Object.prototype.toString.call(date) === '[object Date]'
}

export function formatDate(date: Date) {
  if (!date)
    return ''

  return date.toLocaleDateString()
}

export function randomId(): string {
  return Math.random().toString(36).substring(2) + Math.random().toString(36).substring(2)
}

export interface IOnWheelCallbacks {
   up?: () => void
   down?: () => void
}

export function onWheel(elem: Element, { up = () => { }, down = () => { } }: IOnWheelCallbacks) {
  const onWheelListener: EventListener = e => {
    e = e || window.event

    // wheelDelta не дает возможность узнать количество пикселей
    // @ts-ignore
    const delta = e.deltaY || e.detail || -e.wheelDelta

    if (e.preventDefault)
      e.preventDefault()
    else
      e.returnValue = false

    if (delta < 0)
      up()
    else
      down()
  }

  if (elem.addEventListener) {
    if ('onwheel' in document) {
      // IE9+, FF17+, Ch31+
      elem.addEventListener('wheel', onWheelListener)
    } else if ('onmousewheel' in document) {
      // устаревший вариант события
      elem.addEventListener('mousewheel', onWheelListener)
    } else {
      // Firefox < 17
      elem.addEventListener('MozMousePixelScroll', onWheelListener)
    }
  } else {
    // IE8-
    // @ts-ignore
    elem.attachEvent('onmousewheel', onWheelListener)
  }
}
