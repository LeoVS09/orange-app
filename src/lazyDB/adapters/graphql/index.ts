import { AosSchema } from '@/abstractObjectSchema'
import { GenerateQueryOptions, generateQuery } from './generateQuery'
import { schemaToQueryFields } from './schemaToQuery'

export * from './types'
export * from './schemaToQuery'
export * from './generateQuery'

export interface GenerateQueryFromAosSchemaOptions extends Omit<GenerateQueryOptions, 'fields'> {
    /** fields which will be inside query */
    schema: AosSchema
  }

/** Generate query from AOS schema, for more info check generateQuery function */
export const generateQueryFromAosSchema = ({ operation, schema, variables }: GenerateQueryFromAosSchemaOptions) =>
  generateQuery({
    operation,
    fields: schemaToQueryFields(schema),
    variables
  })
