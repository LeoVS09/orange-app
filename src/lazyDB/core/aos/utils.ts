
const ANY_NON_DIGIT_CHARACTER = /\D/

// TODO: need find better solution for resolve issue

/**
 * Will transform string to number if it possible,
 * actually resolve problem: "When asked item in array by index, index number transform to string",
 * as example: array[3] -> will by catched by proxy as get("3").
 * For resolve this issue, allways tryies cast any string to number.
 * Actaully cannot be casted only for array, because js, under the hood any array actaully is object,
 * and any object with number keys can be interpritated as array. (V8 work this way)
 */
export function toNumberIfCan(name: string | number | symbol): string | number {
  // Symbol just for correct ts typing
  if (typeof name === 'number' || typeof name === 'symbol')
    return name as number

  // Prevent to incorrectly parse id as number
  if (ANY_NON_DIGIT_CHARACTER.test(name))
    return name

  const parsedNumber = Number.parseInt(name, 10)
  if (Number.isInteger(parsedNumber))
    return parsedNumber

  return name
}
