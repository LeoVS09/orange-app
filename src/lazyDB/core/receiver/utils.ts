
export const isPromiseLike = (value: any): value is PromiseLike<any> =>
  typeof value === 'object' && value.then

// TODO: rewrite, code must write without this strange behavior
export function async<T>(value: T | Promise<T>): Promise<T> {
  if (isPromiseLike(value))
    return value

  return Promise.resolve(value)
}
