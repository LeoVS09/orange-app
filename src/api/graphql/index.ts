import * as mutations from './mutations'
import * as queries from './queries'
import {mapRequester} from "./utils";

// TODO: add pre-compiling queries when build project

export const login = mapRequester(mutations.login);

export const register = mapRequester(mutations.register);

export const currentUser = mapRequester(queries.currentUser);

export const searchCountries = mapRequester(queries.searchCountries);

export const countries = mapRequester(queries.countries);

export const problems = mapRequester(queries.problems)

export const problem = mapRequester(queries.problem)
