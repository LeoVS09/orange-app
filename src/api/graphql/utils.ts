import {APIClient, makeClient} from "../graphql/apollo";
import urls from "../urls.json";
import deepMap from "deep-map";

const client = makeClient(urls.DATABASE_SERVER);

// mapRequester transform date string to date objects, but currently TypeScript don't have support
// for so advanced templates, so just know this fact when set types of request result

export function mapRequester<R>(requester: (client: APIClient) => R): R  {
   const wrapped = requester(client)

   // @ts-ignore
   return async (input: any) => {
      // @ts-ignore
      const result = await wrapped(input)
      if(!result)
         return result

      return dateToStringFormatter(result)
   }
}

const keysForMapDays = ['createdAt', 'updatedAt', 'publishedAt', 'publicationDate']

// Transform to date types fields with matched names
function dateToStringFormatter <T>(t: T): T {
   return deepMap<T>(t, (value, key) => {
      if(typeof key === "number")
         return value

      if(keysForMapDays.indexOf(key) === -1)
         return value

      return new Date(value)
   })
}
