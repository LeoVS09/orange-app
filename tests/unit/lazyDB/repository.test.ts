// @ts-ignore next-line
global.fetch = require('node-fetch')
jest.unmock('graphql-tag');
import nock from 'nock'
import { nockAllGraphqlRequests, NockAllGraphqlRequestsControls } from '../utils/graphql.mock'
import cons from '../utils/console.mock'
import { when } from '../utils/when'
import { timeout } from 'rxjs/operators'

// For run tests sequentially (one by one)
// use describe blocks, each describe block runs ony after previus


describe('Repository', () => {
  beforeAll(async () => {
    nock.disableNetConnect()
    cons.mockConsole()
  });

  let requestControls: NockAllGraphqlRequestsControls

  beforeEach(async () => {
    requestControls = await nockAllGraphqlRequests(nock, 'repository')
    // Check graphql query is correct
    requestControls.checkRequestParams = query => 
      expect(query).toMatchSnapshot()
  })

  afterEach(() => {
    nock.cleanAll()
  });

  afterAll(() => {
    cons.restoreConsole()
  })

  // TODO: check order of events
  //  request -> reactive update -> response -> reactive update
  
  // TODO: check display states
  //  isLoading -> before and during request
  //  isChanged -> on change data
  //  isUpdating -> during update
  //  ...

  // TODO: check multiple layers of object
  
  // TODO: test repository without database
  
  it('should generate query for one entity and display', async () => {
    // must be inside test for mock fetch function
    const { Database } = require('@/lazyDb/database/connected/Database')
    const { Repository } = new Database()
    const DataRepository = new Repository('country')

    const reactiveUpdate = jest.fn()
    const node = DataRepository.findOne('test-id', reactiveUpdate) 

    // ask data
    expect(node.name).toBeUndefined()
    expect(node.code).toBeUndefined()
    expect(node.updatedAt).toBeUndefined()
    expect(node.createdAt).toBeUndefined()

    // Wait while lazyDB requested all data and rerender
    await requestControls.waits.wait()
    // one call when requiest start
    await when(() => reactiveUpdate.mock.calls.length > 0)
    expect(reactiveUpdate.mock.calls.length).toBe(1)

    // when data will be displayed
    await when(() => reactiveUpdate.mock.calls.length === 2)
    expect(reactiveUpdate.mock.calls.length).toBe(2)

    expect(node.id).toBe("0SHIyxJBp")
    expect(node.name).toBe("Switzerland")
    expect(node.code).toBe("IN")
    expect(node.updatedAt).toEqual(new Date("2020-04-24T19:29:15+03:00"))
    expect(node.createdAt).toEqual(new Date("2021-04-28T19:29:15+03:00"))

    expect(node).toMatchSnapshot()
  })

  it('should generade query for list by first node and return', async () => {
    // must be inside test for mock fetch function
    const { Database } = require('@/lazyDb/database/connected/Database')
    const { Repository } = new Database()
    const DataRepository = new Repository('country')
  
    const reactiveUpdate = jest.fn()
    const list = DataRepository.list(reactiveUpdate)

    // ask data
    expect(list.nodes[0].name).toBeUndefined()
    expect(list.nodes[0].code).toBeUndefined()
    expect(list.nodes[0].updatedAt).toBeUndefined()
    
    // Wait while lazyDB requested all data and rerender
    await requestControls.waits.wait()
    // only one call, need investigate why, 
    //  possible because debounce is too big
    await when(() => reactiveUpdate.mock.calls.length === 1)
    expect(reactiveUpdate.mock.calls.length).toBe(1)
  
    // when data will be displayed
    // await when(() => reactiveUpdate.mock.calls.length === 2)
    // expect(reactiveUpdate.mock.calls.length).toBe(2)

    expect(list.nodes.length).toBe(15)

    expect(list.nodes[0].id).toBe("gILNkgtqw")
    expect(list.nodes[0].name).toBe("Estonia")
    expect(list.nodes[0].code).toBe("DE")
    expect(list.nodes[0].updatedAt).toEqual(new Date("2020-02-27T00:25:04+03:00"))

    expect(list.nodes[6].id).toBe("vP38-NdB-S")
    expect(list.nodes[6].name).toBe("Guinea-Bissau")
    expect(list.nodes[6].code).toBe("IE")
    expect(list.nodes[6].updatedAt).toEqual(new Date("2020-03-15T00:25:04+03:00"))

    expect(list.nodes).toMatchObject([
      {"id":"gILNkgtqw", "name":"Estonia","code":"DE","updatedAt":new Date("2020-02-27T00:25:04+03:00")},
      {"id":"RvBE9BWZ85","name":"Comoros","code":"CN","updatedAt":new Date("2020-03-02T00:25:04+03:00")},
      {"id":"0Pj8HSsj2e","name":"Mayotte","code":"IN","updatedAt":new Date("2020-03-07T00:25:04+03:00")},
      {"id":"mIBSPRsCW1","name":"Kuwait","code":"US","updatedAt":new Date("2020-03-04T00:25:04+03:00")},
      {"id":"hcQnOCkpmJ","name":"Slovakia (Slovak Republic)","code":"US","updatedAt":new Date("2020-03-07T00:25:04+03:00")},
      {"id":"N8RK5pKUpR","name":"Guinea","code":"FR","updatedAt":new Date("2020-03-12T00:25:04+03:00")},
      {"id":"vP38-NdB-S","name":"Guinea-Bissau","code":"IE","updatedAt":new Date("2020-03-15T00:25:04+03:00")},
      {"id":"HUZ0tFZw41","name":"Nicaragua","code":"MX","updatedAt":new Date("2020-02-29T00:25:04+03:00")},
      {"id":"o_xKHVYYCb","name":"Czech Republic","code":"IE","updatedAt":new Date("2020-03-07T00:25:04+03:00")},
      {"id":"GjmpsFXCnv","name":"India","code":"ES","updatedAt":new Date("2020-02-25T00:25:04+03:00")},
      {"id":"HCGgb2ALkw","name":"Japan","code":"RU","updatedAt":new Date("2020-03-06T00:25:04+03:00")},
      {"id":"G6uw5EaXT8","name":"Monaco","code":"IN","updatedAt":new Date("2020-03-14T00:25:04+03:00")},
      {"id":"tuMpRj14b3","name":"Guadeloupe","code":"MX","updatedAt":new Date("2020-03-09T00:25:04+03:00")},
      {"id":"KF5UhVBanC","name":"Netherlands Antilles","code":"MX","updatedAt":new Date("2020-03-06T00:25:04+03:00")},
      {"id":"FlSxQ0a3nP","name":"Myanmar","code":"FR","updatedAt":new Date("2020-03-09T00:25:04+03:00")}
    ])

    expect(list).toMatchSnapshot()
  })

  it('should generate request for entity and next reqires when more fields need', async () => {
     // must be inside test for mock fetch function
     const { Database } = require('@/lazyDb/database/connected/Database')
     const { Repository } = new Database()
     const DataRepository = new Repository('country')
 
     const reactiveUpdate = jest.fn()
     // need use eindividual id, 
     // for prevent apollo to cache queries
     const node = DataRepository.findOne('test-id2', reactiveUpdate) 

     // ask data
     expect(node.name).toBeUndefined()
     expect(node.updatedAt).toBeUndefined()
     expect(node.createdAt).toBeUndefined()

    // Wait while lazyDB requested all data and rerender
    await requestControls.waits.wait()
    // one call when requiest start
    await when(() => reactiveUpdate.mock.calls.length > 0)
    expect(reactiveUpdate.mock.calls.length).toBe(1)

    // wait when data will be displayed
    // if tests run one by one, requiests may be to fast, and data will be displayed imidiatly
    // await when(() => reactiveUpdate.mock.calls.length === 2)
    // expect(reactiveUpdate.mock.calls.length).toBe(2)

    expect(node.id).toBe('I2o82_J6a')
    expect(node.name).toBe('United Arab Emirates')
    expect(node.updatedAt).toEqual(new Date("2020-05-08T15:31:16.000Z"))
    expect(node.createdAt).toEqual(new Date("2021-01-16T15:31:16.000Z"))

    expect(node).toMatchSnapshot()

    await timeout(1000)
    // ---------------------------------------------------------------
    // Ask additional data
    expect(node.code).toBeUndefined()

    // Wait while lazyDB requested 'code' data and rerender
    await requestControls.waits.wait()

    // next call when requiest end
    // if data not displayed change on 3
    // TODO: investigate this behaivor, probably need add some kind controls to debounce, 
    // for prevent lost steps
    await when(() => reactiveUpdate.mock.calls.length >= 2)
    // expect(reactiveUpdate.mock.calls.length).toBe(2)

    expect(node.id).toBe('I2o82_J6a')
    expect(node.name).toBe('United Arab Emirates')
    expect(node.code).toBe('PT')
    expect(node.updatedAt).toEqual(new Date("2020-05-08T15:31:16.000Z"))
    expect(node.createdAt).toEqual(new Date("2021-01-16T15:31:16.000Z"))

    expect(node).toMatchSnapshot()
  })
  
})
