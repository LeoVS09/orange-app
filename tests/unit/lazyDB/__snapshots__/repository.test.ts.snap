// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Repository should generade query for list by first node and return 1`] = `
Object {
  "query": "query Countries {
  countries {
    totalCount
    nodes {
      id
      nodeId
      name
      code
      updatedAt
      __typename
    }
    __typename
  }
}
",
  "variables": Object {},
}
`;

exports[`Repository should generate query for one entity and display 1`] = `
Object {
  "query": "query Country($id: UUID!) {
  country(id: $id) {
    id
    nodeId
    name
    code
    updatedAt
    createdAt
    __typename
  }
}
",
  "variables": Object {
    "id": "test-id",
  },
}
`;
